smb_control/src/PIDControllerNode.cpp

```diff
@@ -0,0 +1,228 @@
+        {
+            // If no sensors are used, delete the EKF node
+            pure_feed_forward_ = true;
+        }
+    }
+
+    // IMU0 parameters
+    if (use_imu_)
+    {
+        ros::param::set(ns + "/imu0", "/imu");
+        ros::param::set(ns + "/imu0_config", std::vector<bool>{false, false, false, true, true, true, false, false, false, true, true, true, false, false, false});
+        ros::param::set(ns + "/imu0_differential", true);
+        ros::param::set(ns + "/imu0_queue_size", 10);
+        ros::param::set(ns + "/imu0_remove_gravitational_acceleration", true);
+    }
+    else
+    {
+        ros::param::del(ns + "/imu0");
+        ros::param::del(ns + "/imu0_config");
+        ros::param::del(ns + "/imu0_differential");
+        ros::param::del(ns + "/imu0_queue_size");
+        ros::param::del(ns + "/imu0_remove_gravitational_acceleration");
+    }
+
+    // Diagnostic configuration
+    ros::param::set(ns + "/diagnostic_updater/frequency", 1.0);
+    ros::param::set(ns + "/diagnostic_updater/min_frequency", 0.1);
+    ros::param::set(ns + "/diagnostic_updater/max_frequency", 100.0);
+}
+
+void PIDControllerNode::twistCallback(const geometry_msgs::Twist::ConstPtr& msg) {
+    // Get the current angular velocity from the Odom
+    cmd_vel_requested_ = *msg;
+    last_twist_time_ = ros::Time::now();
+}
+
+void PIDControllerNode::odomCallback(const nav_msgs::Odometry::ConstPtr& msg)
+{
+    double angular_velocity = msg->twist.twist.angular.z;
+    double linear_velocity = msg->twist.twist.linear.x; // Assuming x is the forward direction
+
+    // Apply noise threshold
+    if (std::abs(angular_velocity) < 0.01) angular_velocity = 0;
+    if (std::abs(linear_velocity) < 0.01) linear_velocity = 0;
+
+    if (!last_odom_time_.isZero()) { // Check if not the first message
+        dt_ = (msg->header.stamp - last_odom_time_).toSec();
+    }
+    last_odom_time_ = msg->header.stamp; // Update last_odom_time_ for the next call
+
+    // Calculate the error
+    double angular_error = cmd_vel_requested_.angular.z - angular_velocity;
+    double linear_error = cmd_vel_requested_.linear.x - linear_velocity;
+
+    // Reset the integrator if the command should be zero
+    angular_velocity_pid_.resetIfZero(cmd_vel_requested_.angular.z);
+    linear_velocity_pid_.resetIfZero(cmd_vel_requested_.linear.x);
+
+    // Update the PID controller
+
+    cmd_vel_.angular.z = cmd_vel_requested_.angular.z;
+    cmd_vel_.linear.x = cmd_vel_requested_.linear.x;
+
+    if(!pure_feed_forward_)
+    {
+        cmd_vel_.angular.z += angular_velocity_pid_.update(angular_error, dt_);
+        cmd_vel_.linear.x += linear_velocity_pid_.update(linear_error, dt_);
+    }
+
+    // Publish the control output
+    cmd_vel_pub_.publish(cmd_vel_);
+}
+
+void PIDControllerNode::checkTimeout()
+{
+
+    if (ros::Time::now() - last_twist_time_ > ros::Duration(odom_timeout_)) {
+        // No new message within 0.2 seconds, setting cmd_vel_requested_ to zero
```

Comment:
comment is misleading. timeout is a parameter, not fixed

Reply:
Removed the comment, currently its fixly defined as 0.1, I could add it as a parameter if you want? but I think this is quite dangerous to play with. I just added a member variable to make more clear, but its not really nessesary and a good design choice.